storage/src/qmdb/any/db.rs

```diff
@@ -0,0 +1,502 @@
+    /// pruned and those that are not yet committed.
+    pub fn op_count(&self) -> Location {
+        self.log.size()
+    }
+
+    /// Return the inactivity floor location. This is the location before which all operations are
+    /// known to be inactive. Operations before this point can be safely pruned.
+    pub const fn inactivity_floor_loc(&self) -> Location {
+        self.inactivity_floor_loc
+    }
+
+    /// Whether the snapshot currently has no active keys.
+    pub const fn is_empty(&self) -> bool {
+        self.active_keys == 0
+    }
+
+    /// Returns the location of the oldest operation that remains retrievable.
+    pub fn oldest_retained_loc(&self) -> Option<Location> {
+        self.log.oldest_retained_loc()
+    }
+
+    /// Returns the inactivity floor from an authenticated log known to be in a consistent state by
+    /// reading it from the last commit, which is assumed to be the last operation in the log.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the log is empty or the last operation is not a commit floor operation.
+    pub(crate) async fn recover_inactivity_floor(
+        log: &AuthenticatedLog<E, C, H, S>,
+    ) -> Result<Location, Error> {
+        let last_commit_loc = log.size().checked_sub(1).expect("commit should exist");
+        let last_commit = log.read(last_commit_loc).await?;
+        last_commit.has_floor().map_or_else(
+            || unreachable!("last commit is not a CommitFloor operation"),
+            Ok,
+        )
+    }
+
+    /// Get the metadata associated with the last commit.
+    pub async fn get_metadata(&self) -> Result<Option<V::Value>, Error> {
+        match self.log.read(self.last_commit_loc).await? {
+            Operation::CommitFloor(metadata, _) => Ok(metadata),
+            _ => unreachable!("last commit is not a CommitFloor operation"),
+        }
+    }
+}
+
+impl<E, K, V, U, C, I, H> Db<E, C, I, H, U>
+where
+    E: Storage + Clock + Metrics,
+    K: Array,
+    V: ValueEncoding,
+    U: Update<K, V>,
+    C: MutableContiguous<Item = Operation<K, V, U>>,
+    I: UnorderedIndex<Value = Location>,
+    H: Hasher,
+    Operation<K, V, U>: Codec,
+{
+    /// Returns a [Db] initialized from `log`, using `callback` to report snapshot
+    /// building events.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the log is empty or the last operation is not a commit floor operation.
+    pub async fn init_from_log<F>(
+        mut index: I,
+        log: AuthenticatedLog<E, C, H>,
+        known_inactivity_floor: Option<Location>,
+        mut callback: F,
+    ) -> Result<Self, Error>
+    where
+        F: FnMut(bool, Option<Location>),
+    {
+        // If the last-known inactivity floor is behind the current floor, then invoke the callback
+        // appropriately to report the inactive bits.
+        let inactivity_floor_loc = Self::recover_inactivity_floor(&log).await?;
```

Comment:
Do you want to just inline this function here?  I think it's the only remaining callsite.  (I notice it also computes the last_commit_loc, so you can avoid the recomputation on L186).